Mathlib/Topology/Algebra/Monoid.lean

```diff
@@ -393,6 +428,61 @@ end MulOneClass
 
 section ContinuousMul
 
+section Semigroup
+
+variable [TopologicalSpace M] [Semigroup M] [ContinuousMul M]
+
+@[to_additive]
+theorem Subsemigroup.top_closure_mul_self_subset (s : Subsemigroup M) :
+    _root_.closure (s : Set M) * _root_.closure s ⊆ _root_.closure s :=
+  image2_subset_iff.2 fun _ hx _ hy =>
+    map_mem_closure₂ continuous_mul hx hy fun _ ha _ hb => s.mul_mem ha hb
+
+/-- The (topological-space) closure of a subsemigroup of a space `M` with `ContinuousMul` is
+itself a subsemigroup. -/
+@[to_additive "The (topological-space) closure of an additive submonoid of a space `M` with
+`ContinuousAdd` is itself an additive submonoid."]
+def Subsemigroup.topologicalClosure (s : Subsemigroup M) : Subsemigroup M where
+  carrier := _root_.closure (s : Set M)
+  mul_mem' ha hb := s.top_closure_mul_self_subset ⟨_, ha, _, hb, rfl⟩
+
+@[to_additive]
+theorem Subsemigroup.coe_topologicalClosure (s : Subsemigroup M) :
+    (s.topologicalClosure : Set M) = _root_.closure (s : Set M) := rfl
+
+@[to_additive]
+theorem Subsemigroup.le_topologicalClosure (s : Subsemigroup M) : s ≤ s.topologicalClosure :=
+  _root_.subset_closure
+
+@[to_additive]
+theorem Subsemigroup.isClosed_topologicalClosure (s : Subsemigroup M) :
+    IsClosed (s.topologicalClosure : Set M) := isClosed_closure
+
+@[to_additive]
+theorem Subsemigroup.topologicalClosure_minimal (s : Subsemigroup M) {t : Subsemigroup M}
+    (h : s ≤ t) (ht : IsClosed (t : Set M)) : s.topologicalClosure ≤ t := closure_minimal h ht
+
+/-- If a subsemigroup of a topological semigroup is commutative, then so is its topological
+closure. -/
+@[to_additive "If a submonoid of an additive topological monoid is commutative, then so is its
+topological closure."]
+def Subsemigroup.commSemigroupTopologicalClosure [T2Space M] (s : Subsemigroup M)
```

Comment:
Using this, you can replace the proof of [Submonoid.commMonoidTopologicalClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Monoid.html#Submonoid.commMonoidTopologicalClosure) by
```lean
  { s.topologicalClosure.toMonoid, s.toSubsemigroup.commSemigroupTopologicalClosure hs with }
``` 

Could you do that while you're at it?